src/DefaultBuilder/src/WebApplication.cs

```diff
@@ -0,0 +1,243 @@
+            Logger = host.Services.GetRequiredService<ILoggerFactory>().CreateLogger(Environment.ApplicationName);
+        }
+
+        /// <summary>
+        /// The application's configured services.
+        /// </summary>
+        public IServiceProvider Services => _host.Services;
+
+        /// <summary>
+        /// The application's configured <see cref="IConfiguration"/>.
+        /// </summary>
+        public IConfiguration Configuration => _host.Services.GetRequiredService<IConfiguration>();
+
+        /// <summary>
+        /// The application's configured <see cref="IWebHostEnvironment"/>.
+        /// </summary>
+        public IWebHostEnvironment Environment => _host.Services.GetRequiredService<IWebHostEnvironment>();
+
+        /// <summary>
+        /// Allows consumers to be notified of application lifetime events.
+        /// </summary>
+        public IHostApplicationLifetime ApplicationLifetime => _host.Services.GetRequiredService<IHostApplicationLifetime>();
+
+        /// <summary>
+        /// The logger factory for the application.
+        /// </summary>
+        public ILoggerFactory LoggerFactory => _host.Services.GetRequiredService<ILoggerFactory>();
+
+        /// <summary>
+        /// The default logger for the application.
+        /// </summary>
+        public ILogger Logger { get; }
+
+        /// <summary>
+        /// The list of addresses that the HTTP server is bound to.
+        /// </summary>
+        public IEnumerable<string>? Addresses => ServerFeatures.Get<IServerAddressesFeature>()?.Addresses;
+
+        /// <summary>
+        /// A collection of HTTP features of the server.
+        /// </summary>
+        public IFeatureCollection ServerFeatures => _host.Services.GetRequiredService<IServer>().Features;
+
+        IServiceProvider IApplicationBuilder.ApplicationServices
+        {
+            get => ApplicationBuilder.ApplicationServices;
+            set => ApplicationBuilder.ApplicationServices = value;
+        }
+
+        internal IDictionary<string, object?> Properties => ApplicationBuilder.Properties;
+        IDictionary<string, object?> IApplicationBuilder.Properties => Properties;
+
+        internal ICollection<EndpointDataSource> DataSources => _dataSources;
+        ICollection<EndpointDataSource> IEndpointRouteBuilder.DataSources => DataSources;
+
+        internal IEndpointRouteBuilder RouteBuilder
+        {
+            get
+            {
+                Properties.TryGetValue(EndpointRouteBuilder, out var value);
+                return (IEndpointRouteBuilder)value!;
+            }
+        }
+
+        internal ApplicationBuilder ApplicationBuilder { get; }
+
+        IServiceProvider IEndpointRouteBuilder.ServiceProvider => Services;
+
+        /// <summary>
+        /// Sets the URLs the web server will listen on.
+        /// </summary>
+        /// <param name="urls">A set of urls.</param>
+        public void Listen(params string[] urls)
```

Comment:
Many of these APIs need docs and exceptions that clarify required call ordering. E.g. "This must be called before StartAsync". Most of them should also throw if the host has been disposed.

Reply:
Fortunately, this API is being removed. You you want to specify `params string[] urls` now, you have to call an overload of `RunAsync()`.